command/debug.go

```diff
@@ -524,140 +499,225 @@ func (c *DebugCommand) captureStaticTargets(client *api.Client) error {
 		c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "host_info.json", err))
 	}
 
 	return nil
 }
 
-func (c *DebugCommand) intervalCapture(client *api.Client, idxCount int, startTime time.Time, errCh chan<- *captureError) {
-	defer c.pollingWg.Done()
+func (c *DebugCommand) collectHostInfo(ctx context.Context) {
+	idxCount := 0
 
-	var wg sync.WaitGroup
-	currentTimestamp := time.Now().UTC()
+	// intervalTicker needs to be buffered first in order to trigger a run
+	// right away.
+	intervalTicker := make(chan time.Time, 1)
+	intervalTicker <- time.Now()
+	tick := time.NewTicker(c.flagInterval)
+	go func() {
+		for t := range tick.C {
+			intervalTicker <- t
+		}
+	}()
 
-	if strutil.StrListContains(c.flagTargets, "host") {
-		c.logger.Info("capturing host information", "count", idxCount)
+	for {
+		select {
+		case <-ctx.Done():
+			tick.Stop()
+			return
+		case <-intervalTicker:
+			c.logger.Info("capturing host information", "count", idxCount)
+			idxCount++
 
-		wg.Add(1)
-		go func() {
-			r := client.NewRequest("GET", "/v1/sys/host-info")
-			resp, err := client.RawRequest(r)
+			r := c.client.NewRequest("GET", "/v1/sys/host-info")
+			resp, err := c.client.RawRequestWithContext(ctx, r)
 			if err != nil {
-				errCh <- newCaptureError("host", err)
+				c.captureError("host", err)
 			}
 			if resp != nil {
 				defer resp.Body.Close()
 
 				secret, err := api.ParseSecret(resp.Body)
 				if err != nil {
-					errCh <- newCaptureError("host", err)
+					c.captureError("host", err)
 				}
 				if hostEntry := secret.Data; hostEntry != nil {
-					c.hostInfoCh <- hostEntry
+					c.hostInfoCollection = append(c.hostInfoCollection, hostEntry)
 				}
 			}
-			wg.Done()
-		}()
+		}
 	}
+}
 
-	if strutil.StrListContains(c.flagTargets, "pprof") {
-		c.logger.Info("capturing pprof data", "count", idxCount)
+func (c *DebugCommand) collectMetrics(ctx context.Context) {
+	idxCount := 0
+
+	// intervalTicker needs to be buffered first in order to trigger a run
+	// right away.
+	intervalTicker := make(chan time.Time, 1)
+	intervalTicker <- time.Now()
+	tick := time.NewTicker(c.flagMetricsInterval)
+	go func() {
+		for t := range tick.C {
+			intervalTicker <- t
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			tick.Stop()
+			return
+		case <-intervalTicker:
+			c.logger.Info("capturing metrics", "count", idxCount)
+			idxCount++
+
+			healthStatus, err := c.client.Sys().Health()
+			if err != nil {
+				c.captureError("metrics", err)
+				continue
+			}
 
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
+			// Check replication status. We skip on processing metrics if we're one
+			// of the following (since the request will be forwarded):
+			// 1. Any type of DR Node
+			// 2. Non-DR, non-performance standby nodes
+			switch {
+			case healthStatus.ReplicationDRMode == "secondary":
+				c.logger.Info("skipping metrics capture on DR secondary node")
+				continue
```

Comment:
Since this isn't going to change, why not return here instead of continue?

Reply:
If I return here, it means that the rungroup that called this is also going to return, triggering an interrupt and ending the whole capture.

Reply:
The rungroup is only going to end everything if someone returns an error.

Reply:
I'm wrong, thinking of errgroup.  Nvm.